### PR TITLE
factor out AInvOperator in Stokes solver

### PR DESCRIPTION
**Pull request by tjhei**
FYI @quangx

### CODE REVIEW

File: source/simulator/solver.cc [hunk 1]
```diff
@@ -182,19 +182,11 @@ namespace aspect
...
-         *     the inverse of the A block.
          **/
         BlockSchurPreconditioner (const LinearAlgebra::BlockSparseMatrix     &S,
                                   const LinearAlgebra::BlockSparseMatrix     &Spre,
                                   const SInvOperator                         &S_inverse_operator,
-                                  const AInvOperator                         &A_inv_operator,
-                                  const bool                                  do_solve_A,
-                                  const bool                                  A_block_is_symmetric,
-                                  const double                                A_block_tolerance);
+                                  const AInvOperator                         &A_inv_operator);
```
Comment (bangerth) [inline source/simulator/solver.cc:189]: Ah, nice job hiding these details!

File: source/simulator/solver.cc [hunk 2]
```diff
@@ -231,28 +214,20 @@ namespace aspect
...
     {}
 
     template <class AInvOperator, class SInvOperator>
     unsigned int
     BlockSchurPreconditioner<AInvOperator, SInvOperator>::
     n_iterations_A() const
     {
-      return n_iterations_A_;
+      return A_inv_operator.n_iterations;
     }
```
Comment (bangerth) [inline source/simulator/solver.cc:231]: Hm, this is unsatisfying. It requires you to look into the inner object. Is there no better way?

File: source/simulator/solver.cc [hunk 3]
```diff
@@ -287,23 +262,88 @@ namespace aspect
...
+      * as an approximation for the inverse of the velocity (A) block.
+      * This operator can either just apply the preconditioner (AMG)
+      * or perform an inner CG solve with the same preconditioner.
+      * Here, the Schur complement is approximated by
+      * the pressure mass matrix weighted by the inverse of viscosity and
+      * the inverse is computed with a CG solve preconditioned by
+      * PreconditionerMp passed to the constructor.
+      */
+    template <class PreconditionerA>
+    class InverseVelocityBlock
```
Comment (bangerth) [inline source/simulator/solver.cc:280]: The second sentence of the comment is not relevant in the current context.

File: source/simulator/solver.cc [hunk 4]
```diff
@@ -287,23 +262,88 @@ namespace aspect
...
+           * @param do_solve_A A flag indicating whether we should actually solve with
+           *     the matrix $A$, or only apply one preconditioner step with it.
+           * @param A_block_is_symmetric A flag indicating whether the matrix $A$ is symmetric.
+           * @param A_block_tolerance The tolerance for the CG solver which computes
+           *     the inverse of the A block.
+        */
+        InverseVelocityBlock(const TrilinosWrappers::SparseMatrix &matrix,
+                             const PreconditionerA &preconditioner,
+                             const bool do_solve_A,
+                             const bool A_block_is_symmetric,
```
Comment (bangerth) [inline source/simulator/solver.cc:293]: If you do document parameters, document them all :-) Might also be nice to start with a one-sentence summary of what the function does (here, "Constructor.").

**Review by bangerth — COMMENTED** [review]
Right direction. I'm unhappy about exposing the member variable. Perhaps let the class keep a reference to a variable living in an outer scope that keeps track of the number of variables?

I'm bring this up because in the long run, it might actually be nice to replace these classes by just `LinearOperator` objects. We could do away with the template arguments of the class and just pass in generic linear operator objects.

File: source/simulator/solver.cc (hunk 3 above)
Comment (tjhei) [reply source/simulator/solver.cc:280]: Thanks, this was a copy&paste error.

File: source/simulator/solver.cc (hunk 2 above)
Comment (tjhei) [reply source/simulator/solver.cc]: You are of course correct.

**Comment by tjhei** [comment]
fixed, thank you.

File: source/simulator/solver.cc [hunk 5]
```diff
@@ -287,23 +242,88 @@ namespace aspect
...
+                             const double solver_tolerance);
+
+        void vmult(TrilinosWrappers::MPI::Vector &dst,
+                   const TrilinosWrappers::MPI::Vector &src) const;
+
+        unsigned int n_iterations() const;
+
+      private:
+        mutable unsigned int n_iterations_ = 0;
```
Comment (bangerth) [inline source/simulator/solver.cc:281]: Just initialize it in the constructor, along with all the other variables you initialize there.

**Review by bangerth — APPROVED** [review]

**Comment by tjhei** [comment]
updated.

**Comment by tjhei** [comment]
test failure is unrelated, see #5743

**Comment by tjhei** [comment]
I will merge. What is the worst that can happen? :-)

**Pull request merged**